File: find-scholar-links.js
#!/usr/bin/env node
// Find Google Scholar profile links for all PIs
// Searches with "neuro" appended to names

const fs = require('fs');
const OUTPUT_FILE = 'scholar-links.json';

// Firestore REST API
const PROJECT_ID = 'compneuroparis';
const FIRESTORE_URL = `https://firestore.googleapis.com/v1/projects/${PROJECT_ID}/databases/(default)/documents`;

async function fetchAllPIs() {
  const pis = [];
  let pageToken = null;

  while (true) {
    let url = `${FIRESTORE_URL}/groups?pageSize=300`;
    if (pageToken) url += `&pageToken=${pageToken}`;

    const res = await fetch(url);
    const data = await res.json();
    const docs = data.documents || [];

    for (const doc of docs) {
      const fields = doc.fields || {};
      const id = doc.name.split('/').pop();
      const name = fields.name?.stringValue || '';
      const links = (fields.links?.arrayValue?.values || []).map(v => {
        const m = v.mapValue?.fields || {};
        return {
          label: m.label?.stringValue || '',
          url: m.url?.stringValue || ''
        };
      });
      const hasScholar = links.some(l => l.url && l.url.includes('scholar.google'));
      const scholarUrl = links.find(l => l.url && l.url.includes('scholar.google'))?.url || null;

      pis.push({ id, name, hasScholar, existingScholarUrl: scholarUrl });
    }

    if (data.nextPageToken) {
      pageToken = data.nextPageToken;
    } else {
      break;
    }
  }

  return pis;
}

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function searchScholar(name) {
  const query = encodeURIComponent(`${name} neuro`);
  const url = `https://scholar.google.com/citations?view_op=search_authors&mauthors=${query}`;

  const res = await fetch(url, {
    headers: {
      'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
      'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
      'Accept-Language': 'en-US,en;q=0.9',
    }
  });

  if (res.status === 429) {
    return { error: 'rate_limited', url };
  }

  const html = await res.text();

  // Check for CAPTCHA
  if (html.includes('captcha') || html.includes('unusual traffic')) {
    return { error: 'captcha', url };
  }

  // Extract profile links: /citations?user=XXXXX
  const profileRegex = /\/citations\?user=([a-zA-Z0-9_-]+)/g;
  const matches = [];
  let match;
  while ((match = profileRegex.exec(html)) !== null) {
    matches.push(match[1]);
  }

  // Deduplicate
  const uniqueUsers = [...new Set(matches)];

  if (uniqueUsers.length === 0) {
    return { error: 'not_found', url };
  }

  // Extract names associated with each profile for verification
  const results = uniqueUsers.map(userId => ({
    userId,
    profileUrl: `https://scholar.google.com/citations?user=${userId}`
  }));

  return { results, searchUrl: url };
}

async function main() {
  console.log('Fetching all PIs from Firestore...');
  const pis = await fetchAllPIs();
  console.log(`Found ${pis.length} PIs total`);

  const alreadyHave = pis.filter(p => p.hasScholar).length;
  console.log(`${alreadyHave} already have Google Scholar links`);
  console.log(`${pis.length - alreadyHave} need Scholar links\n`);

  // Load existing progress if any
  let results = {};
  if (fs.existsSync(OUTPUT_FILE)) {
    results = JSON.parse(fs.readFileSync(OUTPUT_FILE, 'utf-8'));
    console.log(`Loaded ${Object.keys(results).length} existing results from ${OUTPUT_FILE}\n`);
  }

  let searched = 0;
  let found = 0;
  let notFound = 0;
  let errors = 0;
  let skipped = 0;

  for (const pi of pis) {
    // Skip if already searched
    if (results[pi.id]) {
      skipped++;
      continue;
    }

    // Skip if already has a scholar link in Firestore
    if (pi.hasScholar) {
      results[pi.id] = {
        name: pi.name,
        status: 'already_has',
        scholarUrl: pi.existingScholarUrl
      };
      skipped++;
      continue;
    }

    searched++;
    process.stdout.write(`[${searched}] Searching: ${pi.name}... `);

    const result = await searchScholar(pi.name);

    if (result.error === 'rate_limited' || result.error === 'captcha') {
      console.log(`BLOCKED (${result.error}) - saving progress and stopping`);
      errors++;
      // Save progress before stopping
      fs.writeFileSync(OUTPUT_FILE, JSON.stringify(results, null, 2));
      console.log(`\nSaved progress to ${OUTPUT_FILE}`);
      console.log(`Searched: ${searched}, Found: ${found}, Not found: ${notFound}, Errors: ${errors}`);
      console.log(`\nRe-run the script to continue from where we left off.`);
      process.exit(1);
    }

    if (result.error === 'not_found') {
      console.log('NOT FOUND');
      notFound++;
      results[pi.id] = {
        name: pi.name,
        status: 'not_found',
        searchUrl: result.url
      };
    } else {
      const topResult = result.results[0];
      console.log(`FOUND -> ${topResult.profileUrl} (${result.results.length} results)`);
      found++;
      results[pi.id] = {
        name: pi.name,
        status: 'found',
        scholarUrl: topResult.profileUrl,
        allResults: result.results,
        totalResults: result.results.length,
        searchUrl: result.searchUrl
      };
    }

    // Save after each search
    fs.writeFileSync(OUTPUT_FILE, JSON.stringify(results, null, 2));

    // Rate limit: wait between requests
    await sleep(3000 + Math.random() * 2000); // 3-5 seconds
  }

  console.log(`\n=== SUMMARY ===`);
  console.log(`Total PIs: ${pis.length}`);
  console.log(`Skipped (already done): ${skipped}`);
  console.log(`Searched: ${searched}`);
  console.log(`Found: ${found}`);
  console.log(`Not found: ${notFound}`);
  console.log(`Errors: ${errors}`);
  console.log(`\nResults saved to ${OUTPUT_FILE}`);
}

main().catch(console.error);

File: tools/scholar-links/find-scholar-links.py
#!/usr/bin/env python3
"""Find Google Scholar profile links for all PIs using DuckDuckGo search via curl."""

import json
import re
import time
import subprocess
import urllib.parse
import sys
import os

OUTPUT_FILE = 'scholar-links.json'
PIS_FILE = 'all-pis.json'

def search_ddg(name):
    """Search DuckDuckGo for a PI's Google Scholar profile using curl."""
    query = f'{name} Google Scholar neuroscience'
    url = f'https://html.duckduckgo.com/html/?q={urllib.parse.quote_plus(query)}'

    try:
        result = subprocess.run(
            ['curl', '-s', '-L', '--max-time', '15',
             '-H', 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
             url],
            capture_output=True, text=True, timeout=20
        )
        html = result.stdout
    except Exception as e:
        return {'error': str(e)}

    if not html:
        return {'error': 'empty_response'}

    # Check for rate limiting
    if 'captcha' in html.lower() or 'robot' in html.lower():
        return {'error': 'captcha'}

    # Find all scholar.google.com/citations?user= URLs (may be URL-encoded)
    # First decode any URL encoding
    decoded = urllib.parse.unquote(html)
    user_ids = re.findall(r'scholar\.google\.[a-z.]+/citations\?user=([A-Za-z0-9_-]+)', decoded)
    unique_ids = list(dict.fromkeys(user_ids))

    if not unique_ids:
        return {'error': 'not_found'}

    return {
        'results': [
            {'userId': uid, 'profileUrl': f'https://scholar.google.com/citations?user={uid}'}
            for uid in unique_ids
        ]
    }

def main():
    # Load PIs
    with open(PIS_FILE) as f:
        pis = json.load(f)

    print(f'Total PIs: {len(pis)}')

    # Load existing results
    results = {}
    if os.path.exists(OUTPUT_FILE):
        with open(OUTPUT_FILE) as f:
            results = json.load(f)
        print(f'Loaded {len(results)} existing results')

    # Mark PIs that already have scholar links
    for pi in pis:
        if pi['has_scholar'] and pi['id'] not in results:
            results[pi['id']] = {
                'name': pi['name'],
                'status': 'already_has',
                'scholarUrl': pi['scholar_url']
            }

    need_search = [p for p in pis if p['id'] not in results]
    print(f'Need to search: {len(need_search)}')
    print()

    searched = 0
    found = 0
    not_found = 0
    errors = 0

    for pi in need_search:
        searched += 1
        sys.stdout.write(f'[{searched}/{len(need_search)}] {pi["name"]}... ')
        sys.stdout.flush()

        result = search_ddg(pi['name'])

        if result.get('error') == 'captcha':
            print('CAPTCHA - waiting 60s and retrying...')
            sys.stdout.flush()
            time.sleep(60)
            result = search_ddg(pi['name'])
            if result.get('error') == 'captcha':
                print('CAPTCHA again - saving and stopping')
                with open(OUTPUT_FILE, 'w') as f:
                    json.dump(results, f, indent=2)
                print(f'\nSaved {len(results)} results. Re-run to continue.')
                sys.exit(1)

        if result.get('error') == 'not_found':
            # Try with just "neuro" instead of "neuroscience"
            time.sleep(2)
            result2 = search_ddg_query(f'{pi["name"]} neuro Google Scholar')
            if result2.get('results'):
                result = result2

        if result.get('error'):
            print(f'NOT FOUND ({result["error"]})')
            not_found += 1
            results[pi['id']] = {
                'name': pi['name'],
                'status': 'not_found',
                'error': result['error']
            }
        else:
            top = result['results'][0]
            n_results = len(result['results'])
            print(f'FOUND -> {top["profileUrl"]} ({n_results} result{"s" if n_results > 1 else ""})')
            found += 1
            results[pi['id']] = {
                'name': pi['name'],
                'status': 'found',
                'scholarUrl': top['profileUrl'],
                'allResults': result['results'],
                'totalResults': n_results
            }

        # Save after each search
        with open(OUTPUT_FILE, 'w') as f:
            json.dump(results, f, indent=2)

        # Rate limit: 4-6 seconds between requests
        delay = 4 + (searched % 3)
        time.sleep(delay)

    print(f'\n=== SUMMARY ===')
    print(f'Searched: {searched}')
    print(f'Found: {found}')
    print(f'Not found: {not_found}')
    print(f'Total results saved: {len(results)}')
    print(f'\nResults saved to {OUTPUT_FILE}')


def search_ddg_query(query):
    """Search DuckDuckGo with a custom query string."""
    url = f'https://html.duckduckgo.com/html/?q={urllib.parse.quote_plus(query)}'

    try:
        result = subprocess.run(
            ['curl', '-s', '-L', '--max-time', '15',
             '-H', 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
             url],
            capture_output=True, text=True, timeout=20
        )
        html = result.stdout
    except Exception as e:
        return {'error': str(e)}

    if not html:
        return {'error': 'empty_response'}

    if 'captcha' in html.lower() or 'robot' in html.lower():
        return {'error': 'captcha'}

    decoded = urllib.parse.unquote(html)
    user_ids = re.findall(r'scholar\.google\.[a-z.]+/citations\?user=([A-Za-z0-9_-]+)', decoded)
    unique_ids = list(dict.fromkeys(user_ids))

    if not unique_ids:
        return {'error': 'not_found'}

    return {
        'results': [
            {'userId': uid, 'profileUrl': f'https://scholar.google.com/citations?user={uid}'}
            for uid in unique_ids
        ]
    }


if __name__ == '__main__':
    main()
